detection/meson.build

```diff
@@ -0,0 +1,433 @@
+  cfg_data.set('FLINT_WANT_GMP_INTERNALS', 1,
+    description: 'Define to enable use of GMP internals.')
+endif
+
+if pretty_tests_opt.enabled()
+  cfg_data.set('FLINT_WANT_PRETTY_TESTS', 1,
+    description: 'Define to enable pretty printing for tests.')
+endif
+
+# -----------------------------------------------------------------------------
+#  internal configuration for config.h
+# -----------------------------------------------------------------------------
+
+cfg_data_internal = cfg_data
+
+#cfg_data_internal.set('AC_APPLE_UNIVERSAL_BUILD', 1)
+
+# XXX: Change this when coverage options are added
+#cfg_data_internal.set('FLINT_COVERAGE', 1)
+
+if cc.has_header_symbol('stdlib.h', 'aligned_alloc')
+  cfg_data_internal.set('HAVE_ALIGNED_ALLOC', 1,
+    description: 'Define to 1 if you have the \'aligned_alloc\' function.')
+endif
+
+ac_headers_check = [
+  'alloca.h',
+  'arm_neon.h',
+  'dlfcn.h',
+  'errno.h',
+  'fenv.h',
+  'float.h',
+  'immintrin.h', # XXX: has_header is not equivalent to autotools check...
+  'inttypes.h',
+  'malloc.h',
+  'math.h',
+  'pthread_np.h',
+  'stdarg.h',
+  'stdint.h',
+  'stdio.h',
+  'stdlib.h',
+  'string.h',
+  'strings.h',
+  'sys/param.h',
+  'sys/stat.h',
+  'sys/types.h',
+  'unistd.h',
+  'windows.h',
+]
+
+foreach h : ac_headers_check
+  if cc.has_header(h)
+    cfg_data_internal.set('HAVE_' + h.underscorify().to_upper(), 1,
+      description: f'Define to 1 if you have the <@h@> header file.')
+  endif
+endforeach
+
+# XXX: Do we need to check for _alligned_malloc?
+
+# XXX: Need better checking for LSYM_PREFIX and related assembly options
+cfg_data_internal.set_quoted('LSYM_PREFIX', 'L',
+  description: 'Assembler local label prefix')
+
+cfg_data_internal.set_quoted('LT_OBJDIR', '.libs/',
+  description: 'Define to the sub-directory where libtool stores uninstalled libraries.')
+cfg_data_internal.set_quoted('PACKAGE_BUGREPORT',
+  'https://github.com/flintlib/flint/issues/',
+  description: 'Define to the address where bug reports for this package should be sent.')
+cfg_data_internal.set_quoted('PACKAGE_NAME', 'FLINT',
+  description: 'Define to the full name of this package.')
+cfg_data_internal.set_quoted('PACKAGE_STRING', 'FLINT ' + FLINT_VERSION_FULL,
+  description: 'Define to the full name and version of this package.')
+cfg_data_internal.set_quoted('PACKAGE_TARNAME', 'flint',
+  description: 'Define to the one symbol short name of this package.')
+cfg_data_internal.set_quoted('PACKAGE_URL', 'https://flintlib.org/',
+  description: 'Define to the home page for this package.')
+cfg_data_internal.set_quoted('PACKAGE_VERSION', FLINT_VERSION_FULL,
+  description: 'Define to the version of this package.')
+
+cfg_data_internal.set('STDC_HEADERS', 1,
+  description: '''Define to 1 if all of the C89 standard headers exist (not just the ones
+   required in a freestanding environment). This macro is provided for
+   backward compatibility; new code need not use it.''')
```

Comment:
In the initial attempt to get to a free standing build that does not depend on the `bootstrap.sh` and the autotools `./configure` I tried to reproduce `config.h` and `flint-config.h` as exactly as possible so that I could diff them to see what defines were missing/different. I got that to a good enough state though so what is needed now is just ensuring that the logic for choosing the actual defined values is correct and. The unneeded defines can be removed in later cleanup.

Also it seems that configure generates a bunch of `HAVE_FOO_H` defines that are mostly unused in Flint i.e. these are the only ones actually used:
```console
$ git grep 'HAVE_.*_H'
acinclude.m4:dnl  FLINT_HAVE_FFT_SMALL_ARM_H
acinclude.m4:AC_DEFUN([FLINT_HAVE_FFT_SMALL_ARM_H],
acinclude.m4:dnl  FLINT_HAVE_FFT_SMALL_X86_H
acinclude.m4:AC_DEFUN([FLINT_HAVE_FFT_SMALL_X86_H],
acinclude.m4:AC_REQUIRE([FLINT_HAVE_FFT_SMALL_ARM_H])
acinclude.m4:AC_REQUIRE([FLINT_HAVE_FFT_SMALL_X86_H])
doc/source/history.rst:  * Change ``#ifdef FLINT_HAVE_FFT_SMALL`` to ``#if FLINT_HAVE_FFT_SMALL`` (AA).
src/flint.h.in:#if defined(FLINT_HAVE_VA_LIST) && defined(FLINT_HAVE_FILE)
src/fmpz_lll/is_reduced_d.c:#if HAVE_FENV_H
src/fmpz_lll/is_reduced_d.c:#if HAVE_FENV_H
src/fmpz_lll/is_reduced_d_with_removal.c:#if HAVE_FENV_H
src/fmpz_lll/is_reduced_d_with_removal.c:#if HAVE_FENV_H
src/mpn_extras.h:# define FLINT_HAVE_MUL_FUNC(n, m) FLINT_HAVE_MUL_N_FUNC(n)
src/test/t-io.c:#if HAVE_UNISTD_H && FLINT_COVERAGE
src/thread_pool/init.c:#if FLINT_USES_PTHREAD && defined(HAVE_PTHREAD_NP_H)
src/thread_pool/restore_affinity.c:#if FLINT_USES_PTHREAD && defined(HAVE_PTHREAD_NP_H)
src/thread_pool/set_affinity.c:#if FLINT_USES_PTHREAD && defined(HAVE_PTHREAD_NP_H)
```
Do the others like `HAVE_MALLOC_H` etc still need to be included in `config.h` or `flint-config.h`?

Is it that the configure is just supposed to fail if e.g. `malloc.h` is not available?

Reply:
We require the standard headers `stdio.h`, `stdlib.h`, `stdint.h`, `string.h`, `stdarg.h`, `math.h`, `float.h` and `errno.h`. We check for this during the configuration in order to be able to debug easier.

We assume that `malloc.h` exists for Windows and MinGW, but check if it exists in the configuration as well.

Other things are dependent on other things. `fenv.h` is used if user has it; same goes with `pthread_np.h`.

Reply:
You do not need to generate `HAVE_FOO_H` for everything, that is. However, I want it to be present in any logs generated, so that it is easier to debug.